File: facebook/src/com/facebook/share/internal/ShareContentValidation.java
/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.share.internal;

import android.graphics.Bitmap;
import android.net.Uri;

import com.facebook.FacebookException;
import com.facebook.FacebookSdk;
import com.facebook.internal.Utility;
import com.facebook.internal.Validate;
import com.facebook.share.model.ShareContent;
import com.facebook.share.model.ShareLinkContent;
import com.facebook.share.model.ShareOpenGraphAction;
import com.facebook.share.model.ShareOpenGraphContent;
import com.facebook.share.model.ShareOpenGraphObject;
import com.facebook.share.model.ShareOpenGraphValueContainer;
import com.facebook.share.model.SharePhoto;
import com.facebook.share.model.SharePhotoContent;
import com.facebook.share.model.ShareVideo;
import com.facebook.share.model.ShareVideoContent;

import java.util.List;
import java.util.Locale;
import java.util.Set;

/**
 * com.facebook.share.internal is solely for the use of other packages within the
 * Facebook SDK for Android. Use of any of the classes in this package is
 * unsupported, and they may be modified or removed without warning at any time.
 */
public class ShareContentValidation {

    private static Validator WebShareValidator;
    private static Validator DefaultValidator;
    private static Validator ApiValidator;

    public static void validateForMessage(ShareContent content) {
        validate(content, getDefaultValidator());
    }

    public static void validateForNativeShare(ShareContent content) {
        validate(content, getDefaultValidator());
    }

    public static void validateForWebShare(ShareContent content) {
        validate(content, getWebShareValidator());
    }

    public static void validateForApiShare(ShareContent content) {
        validate(content, getApiValidator());
    }

    private static Validator getDefaultValidator() {
        if (DefaultValidator == null) {
            DefaultValidator = new Validator();
        }
        return DefaultValidator;
    }

    private static Validator getApiValidator() {
        if (ApiValidator == null) {
            ApiValidator = new ApiValidator();
        }
        return ApiValidator;
    }

    private static Validator getWebShareValidator() {
        if (WebShareValidator == null) {
            WebShareValidator = new WebShareValidator();
        }
        return WebShareValidator;
    }

    private static void validate(ShareContent content, Validator validator)
            throws FacebookException {
        if (content == null) {
            throw new FacebookException("Must provide non-null content to share");
        }

        if (content instanceof ShareLinkContent) {
            validator.validate((ShareLinkContent) content);
        } else if (content instanceof SharePhotoContent) {
            validator.validate((SharePhotoContent) content);
        } else if (content instanceof ShareVideoContent) {
            validator.validate((ShareVideoContent) content);
        } else if (content instanceof ShareOpenGraphContent) {
            validator.validate((ShareOpenGraphContent) content);
        }
    }

    private static void validateLinkContent(
            ShareLinkContent linkContent, Validator validator) {
        Uri imageUrl = linkContent.getImageUrl();
        if (imageUrl != null && !Utility.isWebUri(imageUrl)) {
            throw new FacebookException("Image Url must be an http:// or https:// url");
        }
    }

    private static void validatePhotoContent(
            SharePhotoContent photoContent, Validator validator) {
        List<SharePhoto> photos = photoContent.getPhotos();
        if (photos == null || photos.isEmpty()) {
            throw new FacebookException("Must specify at least one Photo in SharePhotoContent.");
        }
        if (photos.size() > ShareConstants.MAXIMUM_PHOTO_COUNT) {
            throw new FacebookException(
                    String.format(
                            Locale.ROOT,
                            "Cannot add more than %d photos.",
                            ShareConstants.MAXIMUM_PHOTO_COUNT));
        }

        for (SharePhoto photo : photos) {
            validator.validate(photo);
        }
    }

    private static void validatePhotoForApi(SharePhoto photo, Validator validator) {
        if (photo == null) {
            throw new FacebookException("Cannot share a null SharePhoto");
        }

        Bitmap photoBitmap = photo.getBitmap();
        Uri photoUri = photo.getImageUrl();

        if (photoBitmap == null) {
            if (photoUri == null) {
                throw new FacebookException(
                        "SharePhoto does not have a Bitmap or ImageUrl specified");
            }

            if (Utility.isWebUri(photoUri) && !validator.isOpenGraphContent()) {
                throw new FacebookException(
                        "Cannot set the ImageUrl of a SharePhoto to the Uri of an image on the " +
                                "web when sharing SharePhotoContent");
            }
        }
    }

    private static void validatePhotoForNativeDialog(SharePhoto photo, Validator validator) {
        validatePhotoForApi(photo, validator);

        if (photo.getBitmap() != null || !Utility.isWebUri(photo.getImageUrl())) {
            Validate.hasContentProvider(FacebookSdk.getApplicationContext());
        }
    }

    private static void validatePhotoForWebDialog(SharePhoto photo, Validator validator) {
        if (photo == null) {
            throw new FacebookException("Cannot share a null SharePhoto");
        }

        Uri imageUri = photo.getImageUrl();
        if (imageUri == null || !Utility.isWebUri(imageUri)) {
            throw new FacebookException(
                    "SharePhoto must have a non-null imageUrl set to the Uri of an image " +
                            "on the web");
        }
    }

    private static void validateVideoContent(
            ShareVideoContent videoContent, Validator validator) {
        validator.validate(videoContent.getVideo());

        SharePhoto previewPhoto = videoContent.getPreviewPhoto();
        if (previewPhoto != null) {
            validator.validate(previewPhoto);
        }
    }

    private static void validateVideo(ShareVideo video, Validator validator) {
        if (video == null) {
            throw new FacebookException("Cannot share a null ShareVideo");
        }

        Uri localUri = video.getLocalUrl();
        if (localUri == null) {
            throw new FacebookException("ShareVideo does not have a LocalUrl specified");
        }

        if (!Utility.isContentUri(localUri) && !Utility.isFileUri(localUri)) {
            throw new FacebookException("ShareVideo must reference a video that is on the device");
        }
    }

    private static void validateOpenGraphContent(
            ShareOpenGraphContent openGraphContent, Validator validator) {
        validator.validate(openGraphContent.getAction());

        String previewPropertyName = openGraphContent.getPreviewPropertyName();
        if (Utility.isNullOrEmpty(previewPropertyName)) {
            throw new FacebookException("Must specify a previewPropertyName.");
        }

        if (openGraphContent.getAction().get(previewPropertyName) == null) {
            throw new FacebookException(
                    "Property \"" + previewPropertyName + "\" was not found on the action. " +
                            "The name of the preview property must match the name of an " +
                            "action property.");
        }
    }

    private static void validateOpenGraphAction(
            ShareOpenGraphAction openGraphAction,
            Validator validator) {
        if (openGraphAction == null) {
            throw new FacebookException("Must specify a non-null ShareOpenGraphAction");
        }

        if (Utility.isNullOrEmpty(openGraphAction.getActionType())) {
            throw new FacebookException("ShareOpenGraphAction must have a non-empty actionType");
        }

        validator.validate((ShareOpenGraphValueContainer) openGraphAction, false);
    }

    private static void validateOpenGraphObject(
            ShareOpenGraphObject openGraphObject,
            Validator validator) {
        if (openGraphObject == null) {
            throw new FacebookException("Cannot share a null ShareOpenGraphObject");
        }

        validator.validate((ShareOpenGraphValueContainer) openGraphObject, true);
    }

    private static void validateOpenGraphValueContainer(
            ShareOpenGraphValueContainer valueContainer,
            Validator validator,
            boolean requireNamespace) {
        Set<String> keySet = valueContainer.keySet();
        for (String key : keySet) {
            validateOpenGraphKey(key, requireNamespace);
            Object o = valueContainer.get(key);
            if (o instanceof List) {
                List list = (List) o;
                for (Object objectInList : list) {
                    if (objectInList == null) {
                        throw new FacebookException(
                                "Cannot put null objects in Lists in " +
                                        "ShareOpenGraphObjects and ShareOpenGraphActions");
                    }
                    validateOpenGraphValueContainerObject(objectInList, validator);
                }
            } else {
                validateOpenGraphValueContainerObject(o, validator);
            }
        }
    }

    private static void validateOpenGraphKey(String key, boolean requireNamespace) {
        if (!requireNamespace) {
            return;
        }

        String[] components = key.split(":");
        if (components.length < 2) {
            throw new FacebookException("Open Graph keys must be namespaced: %s", key);
        }
        for (String component : components) {
            if (component.isEmpty()) {
                throw new FacebookException("Invalid key found in Open Graph dictionary: %s", key);
            }
        }
    }

    private static void validateOpenGraphValueContainerObject(
            Object o, Validator validator) {
        if (o instanceof ShareOpenGraphObject) {
            validator.validate((ShareOpenGraphObject) o);
        } else if (o instanceof SharePhoto) {
            validator.validate((SharePhoto) o);
        }
    }

    private static class WebShareValidator extends Validator {
        @Override
        public void validate(final SharePhotoContent photoContent) {
            throw new FacebookException("Cannot share SharePhotoContent via web sharing dialogs");
        }

        @Override
        public void validate(final ShareVideoContent videoContent) {
            throw new FacebookException("Cannot share ShareVideoContent via web sharing dialogs");
        }

        @Override
        public void validate(final SharePhoto photo) {
            validatePhotoForWebDialog(photo, this);
        }
    }

    private static class ApiValidator extends Validator {
        @Override
        public void validate(final SharePhoto photo) {
            validatePhotoForApi(photo, this);
        }

    }

    private static class Validator {
        private boolean isOpenGraphContent = false;

        public void validate(final ShareLinkContent linkContent) {
            validateLinkContent(linkContent, this);
        }

        public void validate(final SharePhotoContent photoContent) {
            validatePhotoContent(photoContent, this);
        }

        public void validate(final ShareVideoContent videoContent) {
            validateVideoContent(videoContent, this);
        }

        public void validate(final ShareOpenGraphContent openGraphContent) {
            isOpenGraphContent = true;
            validateOpenGraphContent(openGraphContent, this);
        }

        public void validate(final ShareOpenGraphAction openGraphAction) {
            validateOpenGraphAction(openGraphAction, this);
        }

        public void validate(final ShareOpenGraphObject openGraphObject) {
            validateOpenGraphObject(openGraphObject, this);
        }

        public void validate(final ShareOpenGraphValueContainer openGraphValueContainer,
                             boolean requireNamespace) {
            validateOpenGraphValueContainer(openGraphValueContainer, this, requireNamespace);
        }

        public void validate(final SharePhoto photo) {
            validatePhotoForNativeDialog(photo, this);
        }

        public void validate(final ShareVideo video) {
            validateVideo(video, this);
        }

        public boolean isOpenGraphContent() {
            return isOpenGraphContent;
        }
    }
}

File: facebook-common/src/main/java/com/facebook/share/internal/ShareContentValidation.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.share.internal

import com.facebook.FacebookException
import com.facebook.FacebookSdk.getApplicationContext
import com.facebook.internal.Utility.isContentUri
import com.facebook.internal.Utility.isFileUri
import com.facebook.internal.Utility.isNullOrEmpty
import com.facebook.internal.Utility.isWebUri
import com.facebook.internal.Validate.hasContentProvider
import com.facebook.share.model.ShareCameraEffectContent
import com.facebook.share.model.ShareContent
import com.facebook.share.model.ShareLinkContent
import com.facebook.share.model.ShareMedia
import com.facebook.share.model.ShareMediaContent
import com.facebook.share.model.ShareMessengerActionButton
import com.facebook.share.model.ShareMessengerURLActionButton
import com.facebook.share.model.SharePhoto
import com.facebook.share.model.SharePhotoContent
import com.facebook.share.model.ShareStoryContent
import com.facebook.share.model.ShareVideo
import com.facebook.share.model.ShareVideoContent
import java.util.Locale

/**
 * com.facebook.share.internal is solely for the use of other packages within the Facebook SDK for
 * Android. Use of any of the classes in this package is unsupported, and they may be modified or
 * removed without warning at any time.
 */
object ShareContentValidation {
  private val webShareValidator: Validator = WebShareValidator()
  private val defaultValidator: Validator = Validator()
  private val apiValidator: Validator = ApiValidator()
  private val storyValidator: Validator = StoryShareValidator()

  @JvmStatic
  fun validateForMessage(content: ShareContent<*, *>?) {
    validate(content, defaultValidator)
  }

  @JvmStatic
  fun validateForNativeShare(content: ShareContent<*, *>?) {
    validate(content, defaultValidator)
  }

  @JvmStatic
  fun validateForWebShare(content: ShareContent<*, *>?) {
    validate(content, webShareValidator)
  }

  @JvmStatic
  fun validateForApiShare(content: ShareContent<*, *>?) {
    validate(content, apiValidator)
  }

  @JvmStatic
  fun validateForStoryShare(content: ShareContent<*, *>?) {
    validate(content, storyValidator)
  }

  @Throws(FacebookException::class)
  private fun validate(content: ShareContent<*, *>?, validator: Validator) {
    if (content == null) {
      throw FacebookException("Must provide non-null content to share")
    }
    when (content) {
      is ShareLinkContent -> {
        validator.validate(content)
      }
      is SharePhotoContent -> {
        validator.validate(content)
      }
      is ShareVideoContent -> {
        validator.validate(content)
      }
      is ShareMediaContent -> {
        validator.validate(content)
      }
      is ShareCameraEffectContent -> {
        validator.validate(content)
      }
      is ShareStoryContent -> {
        validator.validate(content as ShareStoryContent?)
      }
    }
  }

  private fun validateStoryContent(storyContent: ShareStoryContent?, validator: Validator) {
    if (storyContent == null ||
        (storyContent.backgroundAsset == null && storyContent.stickerAsset == null)) {
      throw FacebookException(
          "Must pass the Facebook app a background asset, a sticker asset, or both")
    }
    if (storyContent.backgroundAsset != null) {
      validator.validate(storyContent.backgroundAsset)
    }
    if (storyContent.stickerAsset != null) {
      validator.validate(storyContent.stickerAsset)
    }
  }

  private fun validateLinkContent(linkContent: ShareLinkContent, validator: Validator) {
    val contentUrl = linkContent.contentUrl
    if (contentUrl != null && !isWebUri(contentUrl)) {
      throw FacebookException("Content Url must be an http:// or https:// url")
    }
  }

  private fun validatePhotoContent(photoContent: SharePhotoContent, validator: Validator) {
    val photos = photoContent.photos
    if (photos == null || photos.isEmpty()) {
      throw FacebookException("Must specify at least one Photo in SharePhotoContent.")
    }
    if (photos.size > ShareConstants.MAXIMUM_PHOTO_COUNT) {
      throw FacebookException(
          String.format(
              Locale.ROOT, "Cannot add more than %d photos.", ShareConstants.MAXIMUM_PHOTO_COUNT))
    }
    for (photo in photos) {
      validator.validate(photo)
    }
  }

  private fun validatePhoto(photo: SharePhoto?) {
    if (photo == null) {
      throw FacebookException("Cannot share a null SharePhoto")
    }
    val photoBitmap = photo.bitmap
    val photoUri = photo.imageUrl
    if (photoBitmap == null && photoUri == null) {
      throw FacebookException("SharePhoto does not have a Bitmap or ImageUrl specified")
    }
  }

  private fun validatePhotoForApi(photo: SharePhoto, validator: Validator) {
    validatePhoto(photo)
    val photoBitmap = photo.bitmap
    val photoUri = photo.imageUrl
    if (photoBitmap == null && isWebUri(photoUri)) {
      throw FacebookException(
          "Cannot set the ImageUrl of a SharePhoto to the Uri of an image on the " +
              "web when sharing SharePhotoContent")
    }
  }

  private fun validatePhotoForNativeDialog(photo: SharePhoto, validator: Validator) {
    validatePhotoForApi(photo, validator)
    if (photo.bitmap != null || !isWebUri(photo.imageUrl)) {
      hasContentProvider(getApplicationContext())
    }
  }

  private fun validatePhotoForWebDialog(photo: SharePhoto, validator: Validator) {
    validatePhoto(photo)
  }

  private fun validateVideoContent(videoContent: ShareVideoContent, validator: Validator) {
    validator.validate(videoContent.video)
    val previewPhoto = videoContent.previewPhoto
    if (previewPhoto != null) {
      validator.validate(previewPhoto)
    }
  }

  private fun validateVideo(video: ShareVideo?, validator: Validator) {
    if (video == null) {
      throw FacebookException("Cannot share a null ShareVideo")
    }
    val localUri =
        video.localUrl ?: throw FacebookException("ShareVideo does not have a LocalUrl specified")
    if (!isContentUri(localUri) && !isFileUri(localUri)) {
      throw FacebookException("ShareVideo must reference a video that is on the device")
    }
  }

  private fun validateMediaContent(mediaContent: ShareMediaContent, validator: Validator) {
    val media = mediaContent.media
    if (media == null || media.isEmpty()) {
      throw FacebookException("Must specify at least one medium in ShareMediaContent.")
    }
    if (media.size > ShareConstants.MAXIMUM_MEDIA_COUNT) {
      throw FacebookException(
          String.format(
              Locale.ROOT, "Cannot add more than %d media.", ShareConstants.MAXIMUM_MEDIA_COUNT))
    }
    for (medium in media) {
      validator.validate(medium)
    }
  }

  @JvmStatic
  fun validateMedium(medium: ShareMedia<*, *>, validator: Validator) {
    when (medium) {
      is SharePhoto -> {
        validator.validate(medium)
      }
      is ShareVideo -> {
        validator.validate(medium as ShareVideo?)
      }
      else -> {
        throw FacebookException(
            String.format(Locale.ROOT, "Invalid media type: %s", medium.javaClass.simpleName))
      }
    }
  }

  private fun validateCameraEffectContent(cameraEffectContent: ShareCameraEffectContent) {
    val effectId = cameraEffectContent.effectId
    if (isNullOrEmpty(effectId)) {
      throw FacebookException("Must specify a non-empty effectId")
    }
  }

  private fun validateShareMessengerActionButton(button: ShareMessengerActionButton?) {
    if (button == null) {
      return
    }
    if (isNullOrEmpty(button.title)) {
      throw FacebookException("Must specify title for ShareMessengerActionButton")
    }
    if (button is ShareMessengerURLActionButton) {
      validateShareMessengerURLActionButton(button)
    }
  }

  private fun validateShareMessengerURLActionButton(button: ShareMessengerURLActionButton) {
    if (button.url == null) {
      throw FacebookException("Must specify url for ShareMessengerURLActionButton")
    }
  }

  private class StoryShareValidator : Validator() {
    override fun validate(storyContent: ShareStoryContent?) {
      validateStoryContent(storyContent, this)
    }
  }

  private class WebShareValidator : Validator() {
    override fun validate(videoContent: ShareVideoContent) {
      throw FacebookException("Cannot share ShareVideoContent via web sharing dialogs")
    }

    override fun validate(mediaContent: ShareMediaContent) {
      throw FacebookException("Cannot share ShareMediaContent via web sharing dialogs")
    }

    override fun validate(photo: SharePhoto) {
      validatePhotoForWebDialog(photo, this)
    }
  }

  private class ApiValidator : Validator() {
    override fun validate(photo: SharePhoto) {
      validatePhotoForApi(photo, this)
    }

    override fun validate(videoContent: ShareVideoContent) {
      if (!isNullOrEmpty(videoContent.placeId)) {
        throw FacebookException("Cannot share video content with place IDs using the share api")
      }
      if (!isNullOrEmpty(videoContent.peopleIds)) {
        throw FacebookException("Cannot share video content with people IDs using the share api")
      }
      if (!isNullOrEmpty(videoContent.ref)) {
        throw FacebookException("Cannot share video content with referrer URL using the share api")
      }
    }

    override fun validate(mediaContent: ShareMediaContent) {
      throw FacebookException("Cannot share ShareMediaContent using the share api")
    }

    override fun validate(linkContent: ShareLinkContent) {
      if (!isNullOrEmpty(linkContent.quote)) {
        throw FacebookException("Cannot share link content with quote using the share api")
      }
    }
  }

  open class Validator {

    open fun validate(linkContent: ShareLinkContent) {
      validateLinkContent(linkContent, this)
    }

    open fun validate(photoContent: SharePhotoContent) {
      validatePhotoContent(photoContent, this)
    }

    open fun validate(videoContent: ShareVideoContent) {
      validateVideoContent(videoContent, this)
    }

    open fun validate(mediaContent: ShareMediaContent) {
      validateMediaContent(mediaContent, this)
    }

    open fun validate(cameraEffectContent: ShareCameraEffectContent) {
      validateCameraEffectContent(cameraEffectContent)
    }

    open fun validate(photo: SharePhoto) {
      validatePhotoForNativeDialog(photo, this)
    }

    open fun validate(video: ShareVideo?) {
      validateVideo(video, this)
    }

    open fun validate(medium: ShareMedia<*, *>) {
      validateMedium(medium, this)
    }

    open fun validate(storyContent: ShareStoryContent?) {
      validateStoryContent(storyContent, this)
    }
  }
}
